clients/pkg/promtail/client/client.go

```diff
@@ -229,6 +241,66 @@ func NewWithTripperware(metrics *Metrics, cfg Config, streamLagLabels []string,
 	return c, nil
 }
 
+// TODO: can this be turned into an implementation of the pkg/ingester/recovery.go Recoverer interface
+// with the current file structure would I need to build a list of all the timestamp/segment files first?
+func (c *client) replayWAL() error {
+	var recordPool = newRecordPool()
+
+	clientBaseWALDir := path.Join(c.cfg.WAL.Dir, c.name)
+	// look for the WAL dir
+	_, err := os.Stat(clientBaseWALDir)
+	if os.IsNotExist(err) {
+		return err
+	}
+	// get tenant directories for the client, since we could have multiple as a result of the tenant pipeline stage
+	// Note: Ignoring errors.
+	matches, _ := filepath.Glob(clientBaseWALDir + "/*")
+	var tenantDirs []string
+	for _, match := range matches {
+		f, _ := os.Stat(match)
+		if f.IsDir() {
+			tenantDirs = append(tenantDirs, match)
+		}
+	}
+	if len(matches) < 1 {
+		return nil
+	}
+	for _, tenantDir := range tenantDirs {
+		tenantID := tenantDir[strings.LastIndex(tenantDir, "/")+1:]
+		r, closer, err := wal.NewWalReader(tenantDir, -1)
+		if err != nil {
+			return err
+		}
+		defer closer.Close()
+
+		// todo, reduce allocations
+		b := newBatch(NoopWAL)
+		seriesRecs := make(map[uint64]model.LabelSet)
+		for r.Next() {
+			rec := recordPool.GetRecord()
+			entry := api.Entry{}
+			if err := ingester.DecodeWALRecord(r.Record(), rec); err != nil {
+				// this error doesn't need to be fatal, we should maybe just throw out this batch?
+			}
+			for _, series := range rec.Series {
+				seriesRecs[uint64(series.Ref)] = util.MapToModelLabelSet(series.Labels.Map())
+			}
+			for _, samples := range rec.RefEntries {
+				if l, ok := seriesRecs[uint64(samples.Ref)]; ok {
+					entry.Labels = l
+					for _, e := range samples.Entries {
+						entry.Entry = e
+						b.replay(entry)
+					}
+
+				}
+			}
+		}
+		c.sendBatch(tenantID, b)
```

Comment:
Not sure if we need to send, the client would most likely do that, every x or when an entry is pushed although this means you need to keep appending to the old wal.

I'm ok to start like this though.

Reply:
Not sure I understand your comment.

Right now writing to the WAL is behind en-queuing a log line within the batch. Here I'm using `batch.replay` to add a sample to the batch without logging it to the WAL, since we're reading an existing WAL segment. And then we're sending the batch. This is essentially set up the same way Prometheus remote write is, we replay the existing WAL before accepting new samples to be sent. The thing I am not doing here yet that I should be is checking the batch size vs the max batch size and sending if we would hit that value. In theory that shouldn't happen right now, since when we swap to a new batch in the happy path we get a new WAL file anyways, but we should be safe.

The thing we should do in the future is decouple the scraping of logs and the sending of them over the client connection by having them be written to the WAL in the middle. Clients would then have tailing code to just tail the WAL segment files.